--- jdk/make/com/sun/script/Makefile	2011-04-20 07:45:26.000000000 +0200
+++ jdk/make/com/sun/script/Makefile	2011-10-27 12:03:42.000000000 +0200
@@ -31,6 +31,8 @@

 AUTO_FILES_JAVA_DIRS = com/sun/script

+OTHER_JAVACFLAGS = -classpath $(RHINO_JAR)
+
 #
 # Files that need to be copied
 #
--- hotspot/src/share/vm/runtime/os.cpp	2012-02-03 10:21:36.000000000 +0100
+++ hotspot/src/share/vm/runtime/os.cpp	2012-02-03 10:33:27.000000000 +0100
@@ -1102,6 +1102,7 @@
         "%/lib/jce.jar:"
         "%/lib/charsets.jar:"
         "%/lib/jfr.jar:"
+        "%/lib/rhino.jar:"
 #ifdef __APPLE__
         "%/lib/JObjC.jar:"
 #endif
--- jdk/make/common/Release.gmk	2012-02-03 10:35:25.000000000 +0100
+++ jdk/make/common/Release.gmk	2012-02-03 10:37:30.000000000 +0100
@@ -821,6 +821,7 @@
 	$(CP) $(RT_JAR) $(JRE_IMAGE_DIR)/lib/rt.jar
 	$(CP) $(RESOURCES_JAR) $(JRE_IMAGE_DIR)/lib/resources.jar
 	$(CP) $(JSSE_JAR) $(JRE_IMAGE_DIR)/lib/jsse.jar
+	$(CP) $(RHINO_JAR) $(JRE_IMAGE_DIR)/lib/rhino.jar
 ifneq ($(JFR_JAR),)
 	$(CP) $(JFR_JAR) $(JRE_IMAGE_DIR)/lib/jfr.jar
 endif
--- jdk/make/com/sun/Makefile	2012-02-03 10:35:25.000000000 +0100
+++ jdk/make/com/sun/Makefile	2012-02-03 10:48:09.000000000 +0100
@@ -31,13 +31,6 @@
 PRODUCT = sun
 include $(BUILDDIR)/common/Defs.gmk
 
-ifndef OPENJDK
-  ORG_EXISTS := $(call DirExists,$(CLOSED_SRC)/share/classes/sun/org,,)
-  ifneq ("$(ORG_EXISTS)", "") 
-    SCRIPT_SUBDIR = script
-  endif
-endif
-
 # jarsigner is part of JRE
 SUBDIRS = java security net/ssl jarsigner
 
