--- jdk/make/sun/font/Makefile.orig	2012-10-12 11:53:52.000000000 +0200
+++ jdk/make/sun/font/Makefile	2012-10-16 17:45:55.000000000 +0200
@@ -133,15 +133,23 @@
     OTHER_LDLIBS += $(FREETYPE_LIB_PATH)/freetype.lib
 else 
   ifeq ($(USING_SYSTEM_FT_LIB), false)
-    FREETYPE_LIB = $(LIB_LOCATION)/$(LIB_PREFIX)freetype.$(LIBRARY_SUFFIX).6
+    ifeq ($(PLATFORM), macosx)
+      FREETYPE_LIB = $(LIB_LOCATION)/$(LIB_PREFIX)freetype.$(LIBRARY_SUFFIX)
+      OTHER_LDLIBS += -lfreetype
+    else
+      FREETYPE_LIB = $(LIB_LOCATION)/$(LIB_PREFIX)freetype.$(LIBRARY_SUFFIX).6
+      OTHER_LDLIBS += -L$(FREETYPE_LIB_PATH) -lfreetype
+    endif
   endif
-  OTHER_LDLIBS += -L$(FREETYPE_LIB_PATH) -lfreetype
 endif 
 
 library:: $(FREETYPE_LIB)
 
 $(FREETYPE_LIB):
 	$(CP) $(FREETYPE_LIB_PATH)/$(LIB_PREFIX)freetype.$(LIBRARY_SUFFIX) $@
+  ifeq ($(PLATFORM), macosx)
+	install_name_tool -id @rpath/libfreetype.dylib $@
+  endif
 	$(install-module-file)
 endif
 
--- jdk/make/common/Defs.gmk.orig	2012-10-16 09:35:50.000000000 +0200
+++ jdk/make/common/Defs.gmk	2012-10-16 09:36:11.000000000 +0200
@@ -180,7 +180,7 @@
   ifdef ALT_FREETYPE_LIB_PATH
     FREETYPE_LIB_PATH = $(ALT_FREETYPE_LIB_PATH)
     ifeq ($(PLATFORM), macosx)
-      USING_SYSTEM_FT_LIB=true
+      USING_SYSTEM_FT_LIB=false
     endif
   else
     ifeq ($(DEVTOOLS_FT_DIR_EXISTS), true)
