diff --git a/build.gradle b/build.gradle
--- a/build.gradle
+++ b/build.gradle
@@ -643,8 +643,13 @@
                 String ver = v.substring(ib + 8, v.size() - 1);
 
                 defineProperty("jdkRuntimeVersion", ver)
-                defineProperty("jdkVersion", jdkRuntimeVersion.split("[_\\-]")[0])
-                defineProperty("jdkBuildNumber", jdkRuntimeVersion.substring(jdkRuntimeVersion.lastIndexOf("-b") + 2))
+		java.util.regex.Matcher m = java.util.regex.Pattern.compile("([^-]*).*-b(\\d+)(?:\$|\\D.*)").matcher(jdkRuntimeVersion);
+		if(m.matches()) {
+                    defineProperty("jdkVersion", m.group(1))
+                    defineProperty("jdkBuildNumber", m.group(2))
+		} else {
+		    fail("Could not parse JDK Version");
+		}
             }
         }
     }
